src/s390x/zipl.rs

```diff
@@ -29,22 +39,159 @@ pub fn chreipl<P: AsRef<Path>>(dev: P) -> Result<()> {
+        let mut input = BufReader::new(file);
+        let entry = NewcBuilder::new(name)
+            .mode(mode as u32)
+            .uid(meta.uid())
+            .gid(meta.gid())
+            .mtime(meta.mtime() as u32)
+            .ino(ino);
+        let size = if mode & 0o40_000 == 0o40_000 {
+            // directory
+            0
+        } else {
+            meta.len()
+        };
+        let mut writer = entry.write(&mut initrd, size as u32);
+        let mut tmp = [0u8; 1024];
+        copy_exactly_n(&mut input, &mut writer, size as u64, &mut tmp)?;
+        writer.finish()?;
+    }
+    // finishing CPIO
+    trailer(&mut initrd)?;
+    initrd.finish()?.flush()?;
+
+    Ok(dest)
+}
+
+fn generate_sdboot<P: AsRef<Path>>(mountpoint: P, boot: P) -> Result<PathBuf> {
     let boot = boot.as_ref();
+    let kernel = Cell::new(String::new());
+    let initrd = Cell::new(String::new());
+    let options = Cell::new(String::new());
 
-    // create dummy config for zipl
-    let mut conffile = Builder::new()
-        .prefix("coreos-installer-zipl.")
+    let read_opts = |contents: &str| {
+        for l in contents.lines() {
+            if let Some(s) = l.strip_prefix("linux ") {
+                kernel.set(s.trim().to_owned());
+            } else if let Some(s) = l.strip_prefix("initrd ") {
+                initrd.set(s.trim().to_owned());
+            } else if let Some(s) = l.strip_prefix("options ") {
+                options.set(s.trim().to_owned());
+            }
+        }
+        Ok(None)
+    };
+    visit_bls_entry(boot, read_opts)?;
+
+    let kernel = kernel.into_inner();
+    let initrd = initrd.into_inner();
+    let options = options.into_inner();
+    if kernel.is_empty() || initrd.is_empty() || options.is_empty() {
+        bail!("bls config doesn't have kernel/initrd/options");
+    }
+    let kernel = format!("{}{}", boot.to_string_lossy(), kernel);
+    let initrd = format!("{}{}", boot.to_string_lossy(), initrd);
+
+    let mut cmdline = Builder::new()
+        .prefix("se-cmdline.")
         .tempfile()
-        .context("creating zipl config")?;
-    let data = format!(
-        "[defaultboot]\ndefaultauto\nprompt=1\ntimeout=5\nsecure=auto\ntarget={}\n",
-        boot.to_str().unwrap()
-    );
-    conffile
-        .write_all(data.as_bytes())
-        .context("writing zipl config")?;
+        .context("creating zipl se cmdline")?;
+    cmdline
+        .write_all(options.as_bytes())
+        .context("writing zipl se cmdline")?;
+
+    let hostkeys = find_files("/etc/se-hostkeys", |p: &PathBuf| {
+        p.file_name()
+            .unwrap_or_default()
+            .to_string_lossy()
+            .starts_with("ibm-z-hostkey-")
+    })?;
+
+    let crypttab = PathBuf::from("/etc/crypttab");
+    let seinitrd = if crypttab.exists() {
+        let mut extra = find_files("/etc/luks/", |p: &PathBuf| p.is_file())?;
+        extra.push(PathBuf::from("/etc/luks")); //add directory to make kernel happy
```

Comment:
Related to the previous comment for example, the `Initrd` API knows to take care of things like this.